File: benchmarks/swift/b13_fuzzy.swift
// Mamdani fuzzy inference system — Fuzzy Systems benchmark (MICAI).
// 3 triangular sets/input, 9-rule base, max-min aggregation, centroid
// defuzzification; only +,-,*,/,min,max so bit-exact. Shared LCG (seed 42).
import Foundation
let Q = 2000000, NP = 100, NSET = 3, NRULE = 9
let SETP: [[Double]] = [[-0.5, 0.0, 0.5], [0.0, 0.5, 1.0], [0.5, 1.0, 1.5]]
var state: UInt64 = 42
@inline(__always) func nextDouble() -> Double {
    state = state &* 6364136223846793005 &+ 1442695040888963407
    return Double(state >> 33) / Double(UInt64(1) << 31)
}
@inline(__always) func tri(_ v: Double, _ a: Double, _ b: Double, _ c: Double) -> Double {
    let left = (v - a) / (b - a), right = (c - v) / (c - b)
    let m = left < right ? left : right
    return m > 0 ? m : 0
}
var zval = [Double](repeating: 0, count: NP)
var os = [Double](repeating: 0, count: NSET * NP)
for j in 0..<NP {
    let z = Double(j) / Double(NP - 1); zval[j] = z
    for s in 0..<NSET { os[s*NP+j] = tri(z, SETP[s][0], SETP[s][1], SETP[s][2]) }
}
var outset = [Int](repeating: 0, count: NRULE)
for xi in 0..<NSET { for yi in 0..<NSET { let sum = xi + yi; outset[xi*NSET+yi] = sum <= 1 ? 0 : (sum == 2 ? 1 : 2) } }
var checksum = 0.0
var muX = [Double](repeating: 0, count: NSET), muY = [Double](repeating: 0, count: NSET)
var fs = [Double](repeating: 0, count: NRULE)
for _ in 0..<Q {
    let x = nextDouble(), y = nextDouble()
    for s in 0..<NSET { muX[s] = tri(x, SETP[s][0], SETP[s][1], SETP[s][2]); muY[s] = tri(y, SETP[s][0], SETP[s][1], SETP[s][2]) }
    for xi in 0..<NSET { for yi in 0..<NSET { let f = muX[xi] < muY[yi] ? muX[xi] : muY[yi]; fs[xi*NSET+yi] = f } }
    var num = 0.0, den = 0.0
    for j in 0..<NP {
        var agg = 0.0
        for r in 0..<NRULE { let osv = os[outset[r]*NP+j]; let m = fs[r] < osv ? fs[r] : osv; if m > agg { agg = m } }
        num += zval[j] * agg; den += agg
    }
    checksum += den > 0 ? num / den : 0
}
print(String(format: "%.6f", checksum))

File: benchmarks/zig/b13_fuzzy.zig
// Mamdani fuzzy inference system — Fuzzy Systems benchmark (MICAI).
// 3 triangular sets/input, 9-rule base, max-min aggregation, centroid
// defuzzification; only +,-,*,/,min,max so bit-exact. Shared LCG (seed 42).
const std = @import("std");
const Q = 2000000;
const NP = 100;
const NSET = 3;
const NRULE = 9;
const setp = [NSET][3]f64{
    .{ -0.5, 0.0, 0.5 },
    .{ 0.0, 0.5, 1.0 },
    .{ 0.5, 1.0, 1.5 },
};
var state: u64 = 42;
fn nextDouble() f64 {
    state = state *% 6364136223846793005 +% 1442695040888963407;
    return @as(f64, @floatFromInt(state >> 33)) / @as(f64, @floatFromInt(@as(u64, 1) << 31));
}
fn tri(v: f64, a: f64, b: f64, c: f64) f64 {
    const left = (v - a) / (b - a);
    const right = (c - v) / (c - b);
    const m = if (left < right) left else right;
    return if (m > 0.0) m else 0.0;
}
var zval: [NP]f64 = undefined;
var os: [NSET * NP]f64 = undefined;
var outset: [NRULE]usize = undefined;
pub fn main(init: std.process.Init) !void {
    const io = init.io;
    for (0..NP) |j| {
        const z = @as(f64, @floatFromInt(j)) / @as(f64, @floatFromInt(NP - 1));
        zval[j] = z;
        for (0..NSET) |s| {
            os[s * NP + j] = tri(z, setp[s][0], setp[s][1], setp[s][2]);
        }
    }
    for (0..NSET) |xi| for (0..NSET) |yi| {
        const sum = xi + yi;
        outset[xi * NSET + yi] = if (sum <= 1) 0 else (if (sum == 2) 1 else 2);
    };
    var checksum: f64 = 0.0;
    var mu_x: [NSET]f64 = undefined;
    var mu_y: [NSET]f64 = undefined;
    var fs: [NRULE]f64 = undefined;
    for (0..Q) |_| {
        const x = nextDouble();
        const y = nextDouble();
        for (0..NSET) |s| {
            mu_x[s] = tri(x, setp[s][0], setp[s][1], setp[s][2]);
            mu_y[s] = tri(y, setp[s][0], setp[s][1], setp[s][2]);
        }
        for (0..NSET) |xi| for (0..NSET) |yi| {
            fs[xi * NSET + yi] = if (mu_x[xi] < mu_y[yi]) mu_x[xi] else mu_y[yi];
        };
        var num: f64 = 0.0;
        var den: f64 = 0.0;
        for (0..NP) |j| {
            var agg: f64 = 0.0;
            for (0..NRULE) |r| {
                const osv = os[outset[r] * NP + j];
                const m = if (fs[r] < osv) fs[r] else osv;
                if (m > agg) agg = m;
            }
            num += zval[j] * agg;
            den += agg;
        }
        checksum += if (den > 0.0) num / den else 0.0;
    }
    var stdout_buffer: [256]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(io, &stdout_buffer);
    const stdout = &stdout_writer.interface;
    try stdout.print("{d:.6}\n", .{checksum});
    try stdout.flush();
}
